src/Bandits/BanditEvaluator.php

```diff
@@ -0,0 +1,268 @@
+     */
+    public static function scoreActions(
+        AttributeSet $subjectAttributes,
+        array $actionsWithContexts,
+        BanditModelData $banditModel
+    ): array {
+        $scores = [];
+        foreach ($actionsWithContexts as $key => $actionAttributes) {
+            if (isset($banditModel->coefficients[$key])) {
+                $scores[$key] = self::scoreAction(
+                    $subjectAttributes,
+                    $actionAttributes,
+                    $banditModel->coefficients[$key]
+                );
+            } else {
+                $scores[$key] = $banditModel->defaultActionScore;
+            }
+        }
+        return $scores;
+    }
+
+    /**
+     * @param AttributeSet $subjectAttributes
+     * @param AttributeSet $actionAttributes
+     * @param ActionCoefficients $coefficients
+     * @return float
+     */
+    private static function scoreAction(
+        AttributeSet $subjectAttributes,
+        AttributeSet $actionAttributes,
+        ActionCoefficients $coefficients
+    ): float {
+        $score = $coefficients->intercept;
+
+        $score += self::scoreNumericAttributes(
+            $coefficients->subjectNumericCoefficients,
+            $subjectAttributes->numericAttributes
+        );
+        $score += self::scoreCategoricalAttributes(
+            $coefficients->subjectCategoricalCoefficients,
+            $subjectAttributes->categoricalAttributes
+        );
+        $score += self::scoreNumericAttributes(
+            $coefficients->actionNumericCoefficients,
+            $actionAttributes->numericAttributes
+        );
+        $score += self::scoreCategoricalAttributes(
+            $coefficients->actionCategoricalCoefficients,
+            $actionAttributes->categoricalAttributes
+        );
+
+        return $score;
+    }
+
+    /**
+     * @param array<string, float> $actionScores
+     * @param float $gamma
+     * @param float $probabilityFloor
+     * @return array<string, float>
+     */
+    public static function weighActions(array $actionScores, float $gamma, float $probabilityFloor): array
+    {
+        $numberOfActions = count($actionScores);
+        $bestActionKey = array_keys($actionScores, max($actionScores))[0];
+
+
+        $minProbability = $probabilityFloor / $numberOfActions;
+
+        $weights = array_filter(
+            $actionScores,
+            function ($key) use ($bestActionKey) {
+                return $key !== $bestActionKey;
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+
+        $bestScore = $actionScores[$bestActionKey];
+        $weights = array_map(
+            function ($score) use ($minProbability, $bestScore, $gamma, $numberOfActions) {
+                return max($minProbability, 1.0 / ($numberOfActions + $gamma * ($bestScore - $score)));
+            },
+            $weights
+        );
+
+        $remainingWeight = max(0.0, 1.0 - array_sum($weights));
+        $weights[$bestActionKey] = $remainingWeight;
```

Comment:
Very clever! Took me a moment to grok; this is adding a new key to `weights` (`bestActionKey` was filtered out above), not overwriting it.